[SonarQube] Shield does not display tech debt

For the net.masterthought:cucumber-reporting project shield does not display tech debt.

Problematic:
https://img.shields.io/sonar/http/nemo.sonarqube.org/net.masterthought:cucumber-reporting/tech_debt.svg
Working:
https://img.shields.io/sonar/http/nemo.sonarqube.org/com.puppycrawl.tools:checkstyle/tech_debt.svg

Project site at Sonar
https://sonarqube.com/dashboard/index?id=net.masterthought%3Acucumber-reporting

I valid in my macbook:
![capture d ecran 2017-04-03 a 23 14 10](https://cloud.githubusercontent.com/assets/5530550/24632338/13998648-18c4-11e7-9d2d-5d209ebc9393.png)

I create a PR: https://github.com/badges/shields/pull/931 but travis-ci is KO with this trace:
gh-badges@1.3.0 lint /home/travis/build/badges/shields
> eslint '**/*.js'
/home/travis/build/badges/shields/server.js
  911:15  error  'apiUrl' is already defined  no-redeclare
  928:19  error  'value' is already defined   no-redeclare
✖ 2 problems (2 errors, 0 warnings)
npm ERR! Linux 4.8.12-040812-generic
npm ERR! argv "/home/travis/.nvm/versions/node/v6.10.1/bin/node" "/home/travis/.nvm/versions/node/v6.10.1/bin/npm" "run" "lint"
npm ERR! node v6.10.1
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! gh-badges@1.3.0 lint: `eslint '**/*.js'`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the gh-badges@1.3.0 lint script 'eslint '**/*.js''.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the gh-badges package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     eslint '**/*.js'
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs gh-badges
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls gh-badges
npm ERR! There is likely additional logging output above.
npm ERR! Please include the following file with any support request:
npm ERR!     /home/travis/build/badges/shields/npm-debug.log
npm ERR! Test failed.  See above for more details.
The command "npm test" exited with 1.

--- a/lib/all-badge-examples.js
+++ b/lib/all-badge-examples.js
@@ -137,11 +137,11 @@ const allBadgeExamples = [
         previewUri: '/sonar/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.svg'
       },
       {
-        title: 'SonarQube Coverage (depreciated API)',
+        title: 'SonarQube Coverage (legacy API)',
         previewUri: '/sonar/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.svg'
       },
       {
-        title: 'SonarQube Tech Debt (depreciated API)',
+        title: 'SonarQube Tech Debt (legacy API)',
         previewUri: '/sonar/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.svg'
       },
       {

--- a/lib/all-badge-examples.js
+++ b/lib/all-badge-examples.js
@@ -130,11 +130,19 @@ const allBadgeExamples = [
       },
       {
         title: 'SonarQube Coverage',
-        previewUri: '/sonar/http/sonar.qatools.ru/ru.yandex.qatools.allure:allure-core/coverage.svg'
+        previewUri: '/sonar/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.svg'
       },
       {
         title: 'SonarQube Tech Debt',
-        previewUri: '/sonar/http/sonar.qatools.ru/ru.yandex.qatools.allure:allure-core/tech_debt.svg'
+        previewUri: '/sonar/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.svg'
+      },
+      {
+        title: 'SonarQube Coverage (depreciated API)',
+        previewUri: '/sonar/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.svg'
+      },
+      {
+        title: 'SonarQube Tech Debt (depreciated API)',
+        previewUri: '/sonar/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.svg'
       },
       {
         title: 'TeamCity CodeBetter Coverage',

--- a/server.js
+++ b/server.js
@@ -827,7 +827,7 @@ cache(function(data, match, sendBadge, request) {
 }));
 
 // SonarQube code coverage
-camp.route(/^\/sonar\/?(.*)?\/(http|https)\/(.*)\/(.*)\/(.*)\.(svg|png|gif|jpg|json)$/,
+camp.route(/^\/sonar\/?([0-9.]+)?\/(http|https)\/(.*)\/(.*)\/(.*)\.(svg|png|gif|jpg|json)$/,
     cache(function(data, match, sendBadge, request) {
       var version = parseFloat(match[1]);
       var scheme = match[2];
@@ -842,7 +842,9 @@ camp.route(/^\/sonar\/?(.*)?\/(http|https)\/(.*)\/(.*)\/(.*)\.(svg|png|gif|jpg|j
         sonarMetricName = 'sqale_debt_ratio';
       }
 
-      var uri = (!!version && version < 5.4) ?
+      const useLegacyApi = !!version && version < 5.4;
+
+      var uri = useLegacyApi ?
           scheme + '://' + serverUrl + '/api/resources?resource=' + buildType + '&depth=0&metrics=' + encodeURIComponent(sonarMetricName) + '&includetrends=true':
           scheme + '://' + serverUrl + '/api/measures/component?componentKey=' + buildType + '&metricKeys=' + encodeURIComponent(sonarMetricName);
 
@@ -869,7 +871,7 @@ camp.route(/^\/sonar\/?(.*)?\/(http|https)\/(.*)\/(.*)\/(.*)\.(svg|png|gif|jpg|j
         try {
           var data = JSON.parse(buffer);
 
-          var value =  parseInt((!!version && version < 5.4) ? data[0].msr[0].val : data.component.measures[0].value);
+          var value =  parseInt(useLegacyApi ? data[0].msr[0].val : data.component.measures[0].value);
 
           if (value === undefined) {
             badgeData.text[1] = 'unknown';

--- a/server.js
+++ b/server.js
@@ -869,7 +869,7 @@ camp.route(/^\/sonar\/?(.*)?\/(http|https)\/(.*)\/(.*)\/(.*)\.(svg|png|gif|jpg|j
         try {
           var data = JSON.parse(buffer);
 
-          var value =  (!!version && version < 5.4) ? data[0].msr[0].val : parseInt(data.component.measures[0].value);
+          var value =  parseInt((!!version && version < 5.4) ? data[0].msr[0].val : data.component.measures[0].value);
 
           if (value === undefined) {
             badgeData.text[1] = 'unknown';

--- a/server.js
+++ b/server.js
@@ -827,13 +827,14 @@ cache(function(data, match, sendBadge, request) {
 }));
 
 // SonarQube code coverage
-camp.route(/^\/sonar\/(http|https)\/(.*)\/(.*)\/(.*)\.(svg|png|gif|jpg|json)$/,
+camp.route(/^\/sonar\/?(.*)?\/(http|https)\/(.*)\/(.*)\/(.*)\.(svg|png|gif|jpg|json)$/,
     cache(function(data, match, sendBadge, request) {
-      var scheme = match[1];
-      var serverUrl = match[2];  // eg, `sonar.qatools.ru`.
-      var buildType = match[3];  // eg, `ru.yandex.qatools.allure:allure-core:master`.
-      var metricName = match[4];
-      var format = match[5];
+      var version = parseFloat(match[1]);
+      var scheme = match[2];
+      var serverUrl = match[3];
+      var buildType = match[4];
+      var metricName = match[5];
+      var format = match[6];
 
       var sonarMetricName = metricName;
       if (metricName === 'tech_debt') {
@@ -841,9 +842,12 @@ camp.route(/^\/sonar\/(http|https)\/(.*)\/(.*)\/(.*)\.(svg|png|gif|jpg|json)$/,
         sonarMetricName = 'sqale_debt_ratio';
       }
 
+      var uri = (!!version && version < 5.4) ?
+          scheme + '://' + serverUrl + '/api/resources?resource=' + buildType + '&depth=0&metrics=' + encodeURIComponent(sonarMetricName) + '&includetrends=true':
+          scheme + '://' + serverUrl + '/api/measures/component?componentKey=' + buildType + '&metricKeys=' + encodeURIComponent(sonarMetricName);
+
       var options = {
-        uri: scheme + '://' + serverUrl + '/api/resources?resource=' + buildType
-          + '&depth=0&metrics=' + encodeURIComponent(sonarMetricName) + '&includetrends=true',
+        uri,
         headers: {
           Accept: 'application/json'
         }
@@ -865,7 +869,7 @@ camp.route(/^\/sonar\/(http|https)\/(.*)\/(.*)\/(.*)\.(svg|png|gif|jpg|json)$/,
         try {
           var data = JSON.parse(buffer);
 
-          var value = data[0].msr[0].val;
+          var value =  (!!version && version < 5.4) ? data[0].msr[0].val : parseInt(data.component.measures[0].value);
 
           if (value === undefined) {
             badgeData.text[1] = 'unknown';

--- a/service-tests/sonarqube.js
+++ b/service-tests/sonarqube.js
@@ -1,0 +1,43 @@
+'use strict';
+
+const Joi = require('joi');
+const ServiceTester = require('./runner/service-tester');
+
+const t = new ServiceTester({ id: 'sonar', title: 'SonarQube' });
+module.exports = t;
+
+t.create('Tech Debt')
+  .get('/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'tech debt',
+    value: Joi.string().regex(/^\d+\%$/)
+  }));
+
+t.create('Coverage')
+  .get('/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'coverage',
+    value: Joi.string().regex(/^\d+\%$/)
+  }));
+
+t.create('Tech Debt (old API supported)')
+  .get('/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'tech debt',
+    value: Joi.string().regex(/^\d+\%$/)
+  }));
+
+t.create('Coverage (old API supported)')
+  .get('/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'coverage',
+    value: Joi.string().regex(/^\d+\%$/)
+  }));
+
+t.create('Tech Debt (old API unsupported)')
+  .get('/4.2/http/sonarqube.com/com.github.dannil:scb-java-client/tech_debt.json')
+  .expectJSON({ name: 'tech debt', value: 'invalid' });
+
+t.create('Coverage (old API unsupported)')
+  .get('/4.2/http/sonarqube.com/com.github.dannil:scb-java-client/coverage.json')
+  .expectJSON({ name: 'coverage', value: 'invalid' });

--- a/service-tests/sonarqube.js
+++ b/service-tests/sonarqube.js
@@ -10,28 +10,28 @@ t.create('Tech Debt')
   .get('/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.json')
   .expectJSONTypes(Joi.object().keys({
     name: 'tech debt',
-    value: Joi.string().regex(/^\d+\%$/)
+    value: Joi.string().regex(/^\d+%$/)
   }));
 
 t.create('Coverage')
   .get('/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.json')
   .expectJSONTypes(Joi.object().keys({
     name: 'coverage',
-    value: Joi.string().regex(/^\d+\%$/)
+    value: Joi.string().regex(/^\d+%$/)
   }));
 
 t.create('Tech Debt (old API supported)')
   .get('/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.json')
   .expectJSONTypes(Joi.object().keys({
     name: 'tech debt',
-    value: Joi.string().regex(/^\d+\%$/)
+    value: Joi.string().regex(/^\d+%$/)
   }));
 
 t.create('Coverage (old API supported)')
   .get('/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.json')
   .expectJSONTypes(Joi.object().keys({
     name: 'coverage',
-    value: Joi.string().regex(/^\d+\%$/)
+    value: Joi.string().regex(/^\d+%$/)
   }));
 
 t.create('Tech Debt (old API unsupported)')

--- a/service-tests/sonarqube.js
+++ b/service-tests/sonarqube.js
@@ -2,6 +2,9 @@
 
 const Joi = require('joi');
 const ServiceTester = require('./runner/service-tester');
+const {
+  isPercentage,
+} = require('./helpers/validators');
 
 const t = new ServiceTester({ id: 'sonar', title: 'SonarQube' });
 module.exports = t;
@@ -10,34 +13,34 @@ t.create('Tech Debt')
   .get('/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.json')
   .expectJSONTypes(Joi.object().keys({
     name: 'tech debt',
-    value: Joi.string().regex(/^\d+%$/)
+    value: isPercentage
   }));
 
 t.create('Coverage')
   .get('/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.json')
   .expectJSONTypes(Joi.object().keys({
     name: 'coverage',
-    value: Joi.string().regex(/^\d+%$/)
+    value: isPercentage
   }));
 
-t.create('Tech Debt (old API supported)')
+t.create('Tech Debt (legacy API supported)')
   .get('/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/tech_debt.json')
   .expectJSONTypes(Joi.object().keys({
     name: 'tech debt',
-    value: Joi.string().regex(/^\d+%$/)
+    value: isPercentage
   }));
 
-t.create('Coverage (old API supported)')
+t.create('Coverage (legacy API supported)')
   .get('/4.2/http/sonar.petalslink.com/org.ow2.petals%3Apetals-se-ase/coverage.json')
   .expectJSONTypes(Joi.object().keys({
     name: 'coverage',
-    value: Joi.string().regex(/^\d+%$/)
+    value: isPercentage
   }));
 
-t.create('Tech Debt (old API unsupported)')
+t.create('Tech Debt (legacy API unsupported)')
   .get('/4.2/http/sonarqube.com/com.github.dannil:scb-java-client/tech_debt.json')
   .expectJSON({ name: 'tech debt', value: 'invalid' });
 
-t.create('Coverage (old API unsupported)')
+t.create('Coverage (legacy API unsupported)')
   .get('/4.2/http/sonarqube.com/com.github.dannil:scb-java-client/coverage.json')
   .expectJSON({ name: 'coverage', value: 'invalid' });

